Mathlib/AlgebraicGeometry/Quasicoherent/Tilde.lean

```diff
@@ -0,0 +1,200 @@
+    refine ⟨Va ⊓ Vb, ⟨ma, mb⟩, Opens.infLELeft _ _ ≫ ia,  sb• ra+ sa•rb , sa * sb, ?_⟩
+    intro y
+    rcases wa (Opens.infLELeft _ _ y : Va) with ⟨nma, wa⟩
+    rcases wb (Opens.infLERight _ _ y : Vb) with ⟨nmb, wb⟩
+    fconstructor
+    · intro H; cases y.1.isPrime.mem_or_mem H <;> contradiction
+    · simp only [Opens.coe_inf, Pi.add_apply, smul_add, map_add, LinearMapClass.map_smul]
+      dsimp at wa wb ⊢
+      rw [← wa, ← wb, ← mul_smul, ← mul_smul]
+      congr 2
+      simp [mul_comm]
+  smul_mem' := by
+    intro r a ha x
+    rcases ha x with ⟨Va, ma, ia, ra, sa, wa⟩
+    rcases r.2 x with ⟨Vr, mr, ir, rr, sr, wr⟩
+    refine ⟨Va ⊓ Vr, ⟨ma, mr⟩, Opens.infLELeft _ _ ≫ ia, rr•ra, sr*sa, ?_⟩
+    intro y
+    rcases wa (Opens.infLELeft _ _ y : Va) with ⟨nma, wa⟩
+    rcases wr (Opens.infLERight _ _ y) with ⟨nmr, wr⟩
+    fconstructor
+    · intro H; cases y.1.isPrime.mem_or_mem H <;> contradiction
+    · simp only [Opens.coe_inf, Pi.smul_apply, LinearMapClass.map_smul]
+      dsimp at wa wr ⊢
+      rw [mul_comm, ← Algebra.smul_def] at wr
+      rw [sections_smul_localizations_def, ← wa, ← mul_smul, ← smul_assoc, mul_comm sr,
+        mul_smul, wr, mul_comm rr, Algebra.smul_def, ← map_mul]
+      rfl
+
+instance (U : (Opens (PrimeSpectrum.Top R))ᵒᵖ) :
+    AddCommGroup ((TildeInType R M).1.obj U) :=
+  inferInstanceAs $ AddCommGroup (sectionsSubmodule R M U)
+
+/--
+`M^~` as a presheaf of abelian groups over `Spec R`
+-/
+def presheafInAddCommGrp : Presheaf AddCommGrp (PrimeSpectrum.Top R) where
+  obj U := AddCommGrp.of ((TildeInType R M).1.obj U)
+  map {U V} i :=
+    { toFun := (TildeInType R M).1.map i
+      map_zero' := rfl
+      map_add' := fun x y => rfl}
+
+/--
+Implementation details:
+checking that after forgeting the abelian group structure of `M^~` as sheaf of abelian groups, we
+get the original sheaf of sets.
+-/
+def presheafCompForget :
+    presheafInAddCommGrp R M ⋙ forget AddCommGrp ≅ (TildeInType R M).1 :=
+  NatIso.ofComponents fun U => Iso.refl _
+
+end Tilde
+
+/--
+`M^~` as a sheaf of abelian groups over `Spec R`
+-/
+def TildeInAddCommGrp : Sheaf AddCommGrp (PrimeSpectrum.Top R) :=
+  ⟨Tilde.presheafInAddCommGrp R M,
+    (TopCat.Presheaf.isSheaf_iff_isSheaf_comp _ _).mpr
+      (TopCat.Presheaf.isSheaf_of_iso (Tilde.presheafCompForget R M).symm (TildeInType R M).cond)⟩
+
+-- `SheafOfModules` want `Sheaf ... RingCat`; but we have a `Sheaf ... CommRingCat`, so we forget.
+local notation3 "𝒪_SpecR" =>
+  sheafCompose (Opens.grothendieckTopology (PrimeSpectrum.Top R))
+    (forget₂ CommRingCat RingCat) |>.obj (Spec.structureSheaf R)
+
+noncomputable instance (U : (Opens (PrimeSpectrum.Top R))ᵒᵖ) :
+    Module ((𝒪_SpecR).val.obj U) ((Tilde.presheafInAddCommGrp R M).obj U) :=
+  inferInstanceAs $ Module _ (Tilde.sectionsSubmodule R M U)
+
+open Tilde in
+/--
+`M^~` as a sheaf of `𝒪_{Spec R}`-modules
+-/
+noncomputable def TildeInModules : SheafOfModules (𝒪_SpecR) where
```

Comment:
You should import `AlgebraicGeometry.Modules.Sheaf` and use `Scheme.Modules`. Also, a better location for this file would be `AlgebraicGeometry.Modules.Tilde`.

Reply:
I wonder if it is better to make "tilde" inside namespace `ModuleCat`, so we can use dot notation. What do others think?

Reply:
`ModuleCat.tilde` seems good to me.